### ISSUE TITLE
ServiceLoaderFeature fails with IllegalArgumentException on abstract classes

### ISSUE THREAD

**Issue by carterkozak** [oracle/graal#3438]
**Describe the issue**
Verified on [GraalVM Community Edition 21.1.0](https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-21.1.0)

The native-image executable fails to produce a binary when ServiceLoaderFeature has not been disabled, and an abstract class  which has a default or no-arg constructor is registered for service loading.

The bug is here: https://github.com/oracle/graal/blob/e0b6ed270fe7f339a5ccfa540123d69d600baf7c/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ServiceLoaderFeature.java#L310-L330

Which reimplements part, but not enough, of the method it delegates to, `registerForReflectiveInstantiation`:

https://github.com/oracle/graal/blob/e0b6ed270fe7f339a5ccfa540123d69d600baf7c/sdk/src/org.graalvm.nativeimage/src/org/graalvm/nativeimage/hosted/RuntimeReflection.java#L125-L138

Much like the case in which a default constructor cannot be found, this case should probably also log and continue.

I believe the root cause is an interpretation of ServiceLoader which is no longer accurate:
The comment in ServiceLoaderFeature incorrectly states:
https://github.com/oracle/graal/blob/e0b6ed270fe7f339a5ccfa540123d69d600baf7c/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ServiceLoaderFeature.java#L312-L317

The comment was correct in java 8, however it's no longer true in java 9+ due to the addition of the `ServiceLoader.Provider<T>` which provides an accessor for the `Class<? extends T>` registered class instance. We don't necessarily know how that class will be used, however the class itself should be reflectively accessible, and considered reachable if the service loader is used.
See both:
* https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.html#stream()
* https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.Provider.html

**Steps to reproduce the issue**
Please include both build steps as well as run steps
1. Create a project which uses java ServiceLoader
2. Register an abstract class which does not define constructors in the relevant `META-INF/services/` file.

I have created a reproducer here: https://github.com/carterkozak/graalvm-serviceloader-reproducer

**Describe GraalVM and your environment:**
 - GraalVM version: Verified on [GraalVM Community Edition 21.1.0](https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-21.1.0)
 - JDK major version: `11`
 - OS: Ubuntu 20.04
 - Architecture: AMD64

**More details**
```
Fatal error:java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Class <ABSTRACT IMPLEMENTATION FQCN> cannot be instantiated reflectively. It must be a non-abstract instance type.
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:600)
        at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
        at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:499)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:370)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:531)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:119)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:568)
Caused by: java.lang.IllegalArgumentException: Class <ABSTRACT IMPLEMENTATION FQCN> cannot be instantiated reflectively. It must be a non-abstract instance type.
        at org.graalvm.sdk/org.graalvm.nativeimage.hosted.RuntimeReflection.registerForReflectiveInstantiation(RuntimeReflection.java:135)
        at com.oracle.svm.hosted.ServiceLoaderFeature.handleType(ServiceLoaderFeature.java:324)
        at com.oracle.svm.hosted.ServiceLoaderFeature.duringAnalysis(ServiceLoaderFeature.java:178)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$14(NativeImageGenerator.java:765)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:71)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:765)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:582)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:495)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Error: Image build request failed with exit status 1
```

**Comment by mcraj017** [comment]
@carterkozak thanks for reporting the issue
we will check it out and get back to you

**Comment by mcraj017** [comment]
@carterkozak please try latest version 21.2.0
 https://github.com/graalvm/graalvm-ce-dev-builds/releases/tag/21.2.0-dev-20210601_2247

**Comment by carterkozak** [comment]
@mcraj017 Still appears to be broken.

**Comment by mcraj017** [comment]
@carterkozak thanks for checking
please provide sample code to reproduce this issue

**Comment by carterkozak** [comment]
@mcraj017 I've created and documented a reproducer here: https://github.com/carterkozak/graalvm-serviceloader-reproducer

Please let me know if you have any questions!

**Comment by mcraj017** [comment]
@carterkozak thanks for that, I will check it out and let you know

**Comment by zapster** [comment]
Hi @carterkozak and thanks for the report, the reproducer and the initial analysis! Very much appreciated.

The issues has been fixed in #3723.

**Comment by carterkozak** [comment]
Wonderful, thank you!
